### DISCUSSION TITLE
Recommended way to build custom DB 19c images

### DISCUSSION THREAD

**Discussion by Champak-Netcracker** [oracle/docker-images#1998] [Q&A]
Folks- I am a Technical Architect from Netcracker Technology, we are looking to build Oracle DB 19c images to run on Kubernetes[Single instance], however, we have mandatory security policies, patches and would like to build our own images from a baseline version, so, two queries:

a) Which is the best place to download the latest baseline images for 19c EE?
b) please share the recommended procedure to build these images?

Thanks/Champak

**Comment by gvenzl** [comment]
Hi @Champak-Netcracker,

The baseline for patched images is the GA version of Oracle 19c EE, which is 19.3.0.

We provide scripts in this repository that allow you to build patched images, see: https://github.com/oracle/docker-images/tree/main/OracleDatabase/SingleInstance/extensions/patching

So the procedure would be:

a) build the base 19.3.0 image
b) use the scripts to derive a patched image from the base image

@yunus-qureshi, @abhisbyk, do you have any other recommendations?

**Reply by Champak-Netcracker** [reply]
Hi @gvenzl , many thanks for your responses, will await replies from @yunus-qureshi @abhisbyk ...additionally, I am looking for answers below:
1) you mentioned the baseline images for 19.3.0 are already patched, does this include all patches, one off patches also suggested for some customers? is there a way to validate this for specific patches[example: the ones suggested for Netcracker centric environment]?
2) Please let me know where I can get the baseline patches images, GA version for 19.8.0
3) If Oracle provided baseline images are already patched, what are the possible reasons or use cases to build our own images?

**Reply by abhisbyk** [reply]
Hi @Champak-Netcracker,
I would like to mention that 19.3.0 baseline image is **not already patched**. What @gvenzl wanted to say is that GA version of Oracle Database 19c EE is the version 19.3.0. This version can be considered as the baseline and used to create patched images by applying patches to it.

Further, to create 19.3.0 baseline image, you need to download the required binaries from the [Oracle Technology Network](https://www.oracle.com/database/technologies/oracle-database-software-downloads.html), place them to `dockerfiles/version` directory and use `buildContainerImage.sh `script to build the image (Detailed instructions are provided [here](https://github.com/oracle/docker-images/blob/main/OracleDatabase/SingleInstance/README.md#building-oracle-database-container-images) ).

After creating this baseline image, you can patch it further (like to version 19.8.0 in your case) by using our patching extension. Instructions to patch the image are [here](https://github.com/oracle/docker-images/blob/main/OracleDatabase/SingleInstance/extensions/README.md). You need to download the required patches from [My Oracle Support](https://support.oracle.com/CSP/ui/flash.html) and place them appropriately as per the [instructions](https://github.com/oracle/docker-images/blob/main/OracleDatabase/SingleInstance/extensions/patching/README.md).

Hope this helps, thanks.

**Reply by Champak-Netcracker** [reply]
Hi @abhisbyk, Thanks for your responses.

Couple of queries:
a) No baseline 19.8.0 image available ? please confirm?
b) All baseline versions are not patched? please confirm? Do we get have an already patched baseline version?
c) Is there a way to validate if a particular image contains a specific patch? please confirm?

/Champak

**Reply by Champak-Netcracker** [reply]
Hi @abhisbyk, Thanks for your responses.

Hi @abhisbyk, adding another query, please share your responses:

a) No baseline 19.8.0 image available ? please confirm?
b) All baseline versions are not patched? please confirm? Do we get have an already patched baseline version?
c) Is there a way to validate if a particular image contains a specific patch? please confirm?
d) for lifecycle and administrative operations, please any documentation available?

/Champak

**Reply by Champak-Netcracker** [reply]
Hi @gvenzl , not sure if @abhisbyk is on vacation, please help with the below queries:

a) No baseline 19.8.0 image available ? please confirm?
b) All baseline versions are not patched? please confirm? Do we get have an already patched baseline version?
c) Is there a way to validate if a particular image contains a specific patch? please confirm?
d) for lifecycle and administrative operations, please any documentation available?

/Champak

**Reply by abhisbyk** [reply]
Hi @Champak-Netcracker ,
Sorry for delayed response. Answers to your queries are as follows:
- Baseline version for 19.8.0 is not available. 
- Also, all baseline versions are not patched by default. We have to manually patch the baseline 19.3.0 image as described in my previous response.
- To validate a particular image contains a specific patch or not, spin up a container with the image and then exec into it. After this, OPatch utility can be used for checking which patches are applied. The command is `$ORACLE_HOME/OPatch/opatch lspatches`.
- For performing lifecycle and admin operations, one needs to exec into the pod and do it. Link for the documentation is [https://docs.oracle.com/en/database/oracle/oracle-database/](https://docs.oracle.com/en/database/oracle/oracle-database/). 

Thanks

**Reply by Champak-Netcracker** [reply]
Many thanks @abhisbyk @gvenzl , I am satisfied with the response, we will get going and come back if required, we can mark this thread as answered.
